rust/agents/watcher/src/watcher.rs

```diff
@@ -583,27 +564,32 @@ impl NomadAgent for Watcher {
                     self.shutdown().await;
                 },
                 improper_res = improper_update_watch_task => {
-                    let state = improper_res??;
-                    if state == State::Failed {
-                        tracing::error!(
-                            "Improper update detected! Notifying all contracts and unenrolling replicas!",
-                        );
-
-                        self.handle_improper_update_failure()
-                            .await
-                            .iter()
-                            .for_each(|res| tracing::info!("{:#?}", res));
-
-                        bail!(
-                            r#"
-                            Improper update detected!
-                            Replicas unenrolled!
-                            Watcher has been shut down!
-                        "#
-                        )
+                    let some_base_error: BaseError = improper_res? // Result<(), E>, where E is either BaseError or Some RPC error
+                        .err() // Option<E>, where E is either BaseError or Some RPC error
+                        .expect("Expected some error on resolve of self.watch_home_fail(...)")
```

Comment:
Should we check if this is `Ok` vs `Err`? I get that the watch task should never return Ok(()) and will only exit on failure but just wasn't sure if that will always be the case. Probably fine just wondering if any thoughts here

Reply:
I have quite some thoughts pro and con this, and in general it should be checked, on the other had if the method only throws an Error, I think it is too much covering all scenarios. Same time if the code on the receiver side says it is possible - it should be covering. 

Either way, ideally there should be no such places which would be causing ambiguity and type abuse. I guess there will be a refactor soon (James' new architecture) and for the clean code it would be good to take into consideration such places of ambiguity.

I push now the change which address this, as it is a fair point to cover Ok(()) anyways